### PR TITLE
fixed 2 memory leaks

### PR DESCRIPTION
**Pull request by heejaechang**
first one is preview workspace leak due to diagnostic service hold onto workspace.
second is capturing project in diagnostic exception handler

### CODE REVIEW

**Comment by heejaechang** [comment]
test vsi please

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs [hunk 1]
```diff
@@ -314,17 +321,35 @@ private void AssertIfNull(ImmutableArray<DiagnosticData> diagnostics)
...
+                    AssertIfNull(data.Workspace);
+                }
+            }
+        }
+
         private struct Data : IEquatable<Data>
         {
-            public readonly Workspace Workspace;
+            private static readonly ConditionalWeakTable<Workspace, WeakReference<Workspace>> s_cache =
+                new ConditionalWeakTable<Workspace, WeakReference<Workspace>>();
```
Comment (Pilchie) [inline src/Features/Core/Portable/Diagnostics/DiagnosticService.cs]: What is the purpose of this?  To reduce allocations of WeakReference objects?

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs [hunk 2]
```diff
@@ -314,17 +321,35 @@ private void AssertIfNull(ImmutableArray<DiagnosticData> diagnostics)
         }
 
         [Conditional("DEBUG")]
-        private void AssertIfNull(DiagnosticData diagnostic)
+        private void AssertIfNull<T>(T diagnostic) where T : class
```
Comment (Pilchie) [inline src/Features/Core/Portable/Diagnostics/DiagnosticService.cs]: rename `diagnostic` now.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs [hunk 3]
```diff
@@ -235,6 +235,7 @@ private void ClearCompilationScopeMap(AnalyzerAndOptions analyzerAndOptions, Ana
                 });
 
                 // Subscribe for exceptions from lazily evaluated localizable strings in the descriptors.
+                // PREVIEW: when are these handle unsubscribe from description?
```
Comment (Pilchie) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs]: `REVIEW` not `PREVIEW`.  Also, I would word this as "When do these unsubscribe from description?"

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs (hunk 2 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/DiagnosticService.cs]: done

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs (hunk 1 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/DiagnosticService.cs]: yes.

**Comment by heejaechang** [comment]
test vsi please

**Pull request closed without merging**